File: ios/Plugin/RevenueCatUIPlugin.swift
import Foundation
import Capacitor
import RevenueCatUI
import PurchasesHybridCommonUI

@objc(RevenueCatUIPlugin)
public class RevenueCatUIPlugin: CAPPlugin {
    
    @objc func presentPaywall(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            let args = self.parsePaywallArgs(from: call)
            
            RevenueCatUI.presentPaywall(
                offeringIdentifier: args.offeringIdentifier,
                requiredEntitlementIdentifier: args.requiredEntitlementIdentifier,
                overrides: args.overrides,
                scene: self.bridge?.viewController) { result in
                    call.resolve(self.mapPaywallResult(result))
                }
        }
    }
    
    @objc func presentPaywallIfNeeded(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            let args = self.parsePaywallArgs(from: call)
            
            // Ensure we have a required entitlement identifier
            guard let requiredEntitlementIdentifier = args.requiredEntitlementIdentifier else {
                call.reject("requiredEntitlementIdentifier is required for presentPaywallIfNeeded")
                return
            }
            
            RevenueCatUI.presentPaywallIfNeeded(
                requiredEntitlementIdentifier: requiredEntitlementIdentifier,
                offeringIdentifier: args.offeringIdentifier,
                overrides: args.overrides,
                scene: self.bridge?.viewController) { result in
                    if let result = result {
                        call.resolve(self.mapPaywallResult(result))
                    } else {
                        call.resolve(nil)
                    }
                }
        }
    }
    
    @objc func presentCustomerCenter(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            RevenueCatUI.presentCustomerCenter(
                parent: self.bridge?.viewController,
                onSuccess: {
                    call.resolve()
                },
                onError: { error in
                    call.reject(error.localizedDescription)
                }
            )
        }
    }
    
    @objc func setPaywallOverrides(_ call: CAPPluginCall) {
        if let overridesDict = call.getObject("overrides") {
            let overrides = self.parseOverrides(from: overridesDict)
            RevenueCatUI.setPaywallOverrides(overrides)
            call.resolve()
        } else {
            call.reject("Missing overrides parameter")
        }
    }
    
    // MARK: - Helper Methods
    
    private func parsePaywallArgs(from call: CAPPluginCall) -> (offeringIdentifier: String?, requiredEntitlementIdentifier: String?, overrides: RevenueCatUI.PaywallOverrides?) {
        let offeringIdentifier = call.getString("offeringIdentifier")
        let requiredEntitlementIdentifier = call.getString("requiredEntitlementIdentifier")
        
        var overrides: RevenueCatUI.PaywallOverrides? = nil
        if let overridesDict = call.getObject("overrides") {
            overrides = parseOverrides(from: overridesDict)
        }
        
        return (offeringIdentifier, requiredEntitlementIdentifier, overrides)
    }
    
    private func parseOverrides(from dict: [String: Any]) -> RevenueCatUI.PaywallOverrides {
        let overrides = RevenueCatUI.PaywallOverrides()
        
        // Font options
        if let fontFamily = dict["fontFamily"] as? String {
            overrides.fontName = fontFamily
        }
        if let titleFontSize = dict["titleFontSize"] as? NSNumber {
            overrides.titleFontSize = titleFontSize.floatValue
        }
        if let bodyFontSize = dict["bodyFontSize"] as? NSNumber {
            overrides.bodyFontSize = bodyFontSize.floatValue
        }
        if let captionFontSize = dict["captionFontSize"] as? NSNumber {
            overrides.captionFontSize = captionFontSize.floatValue
        }
        
        // Colors
        if let backgroundColor = dict["backgroundColor"] as? String {
            overrides.backgroundColor = UIColor(hexString: backgroundColor)
        }
        if let primaryTextColor = dict["primaryTextColor"] as? String {
            overrides.primaryTextColor = UIColor(hexString: primaryTextColor)
        }
        if let secondaryTextColor = dict["secondaryTextColor"] as? String {
            overrides.secondaryTextColor = UIColor(hexString: secondaryTextColor)
        }
        if let accentColor = dict["accentColor"] as? String {
            overrides.accentColor = UIColor(hexString: accentColor)
        }
        
        // Button options
        if let primaryButtonBackgroundColor = dict["primaryButtonBackgroundColor"] as? String {
            overrides.primaryButtonBackgroundColor = UIColor(hexString: primaryButtonBackgroundColor)
        }
        if let primaryButtonTextColor = dict["primaryButtonTextColor"] as? String {
            overrides.primaryButtonTextColor = UIColor(hexString: primaryButtonTextColor)
        }
        if let secondaryButtonBackgroundColor = dict["secondaryButtonBackgroundColor"] as? String {
            overrides.secondaryButtonBackgroundColor = UIColor(hexString: secondaryButtonBackgroundColor)
        }
        if let secondaryButtonTextColor = dict["secondaryButtonTextColor"] as? String {
            overrides.secondaryButtonTextColor = UIColor(hexString: secondaryButtonTextColor)
        }
        
        // Text overrides
        if let title = dict["title"] as? String {
            overrides.title = title
        }
        if let subtitle = dict["subtitle"] as? String {
            overrides.subtitle = subtitle
        }
        if let callToAction = dict["callToAction"] as? String {
            overrides.callToAction = callToAction
        }
        if let restoreButtonText = dict["restoreButtonText"] as? String {
            overrides.restoreButtonText = restoreButtonText
        }
        
        // Display options
        if let displayCloseButton = dict["displayCloseButton"] as? Bool {
            overrides.displayCloseButton = displayCloseButton
        }
        
        return overrides
    }
    
    private func mapPaywallResult(_ result: RevenueCatUI.PaywallResult) -> [String: Any] {
        var resultDict: [String: Any] = [:]
        
        switch result {
        case .purchased(let transactionId):
            resultDict["status"] = "purchased"
            resultDict["transactionId"] = transactionId
            
        case .restored(let transactionId):
            resultDict["status"] = "restored"
            if let transactionId = transactionId {
                resultDict["transactionId"] = transactionId
            }
            
        case .cancelled:
            resultDict["status"] = "cancelled"
            
        case .error(let error):
            resultDict["status"] = "error"
            resultDict["error"] = [
                "code": "\(error.code)",
                "message": error.localizedDescription
            ]
        }
        
        return resultDict
    }
}

// MARK: - UIColor extension for hex string
extension UIColor {
    convenience init(hexString: String) {
        let hex = hexString.trimmingCharacters(in: CharacterSet.alphanumerics.inverted)
        var int = UInt64()
        Scanner(string: hex).scanHexInt64(&int)
        let a, r, g, b: UInt64
        switch hex.count {
        case 3: // RGB (12-bit)
            (a, r, g, b) = (255, (int >> 8) * 17, (int >> 4 & 0xF) * 17, (int & 0xF) * 17)
        case 6: // RGB (24-bit)
            (a, r, g, b) = (255, int >> 16, int >> 8 & 0xFF, int & 0xFF)
        case 8: // ARGB (32-bit)
            (a, r, g, b) = (int >> 24, int >> 16 & 0xFF, int >> 8 & 0xFF, int & 0xFF)
        default:
            (a, r, g, b) = (255, 0, 0, 0)
        }
        self.init(red: CGFloat(r) / 255, green: CGFloat(g) / 255, blue: CGFloat(b) / 255, alpha: CGFloat(a) / 255)
    }
} 

File: purchases-capacitor-ui/android/src/main/java/com/revenuecat/purchases/capacitor/ui/RevenueCatUIPlugin.kt
package com.revenuecat.purchases.capacitor.ui

import android.graphics.Color
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import org.json.JSONObject
import com.revenuecat.purchases.ui.paywall.PaywallActivityLauncher
import com.revenuecat.purchases.ui.paywall.PaywallColor
import com.revenuecat.purchases.ui.paywall.PaywallDialog
import com.revenuecat.purchases.ui.paywall.PaywallFooterColor
import com.revenuecat.purchases.ui.paywall.PaywallFooterView
import com.revenuecat.purchases.ui.paywall.PaywallOptions
import com.revenuecat.purchases.ui.paywall.PaywallOverrides
import com.revenuecat.purchases.ui.paywall.PaywallText
import com.revenuecat.purchases.ui.revenuecatui.CustomerCenter
import com.revenuecat.purchases.ui.revenuecatui.CustomerCenterLauncher

@CapacitorPlugin(name = "RevenueCatUI")
class RevenueCatUIPlugin : Plugin() {

    @PluginMethod
    fun presentPaywall(call: PluginCall) {
        val args = parsePaywallArgs(call)
        
        activity?.let { activity ->
            PaywallActivityLauncher.launch(
                activity = activity,
                offeringIdentifier = args.offeringIdentifier,
                requiredEntitlementIdentifier = args.requiredEntitlementIdentifier,
                overrides = args.overrides,
                listener = CapacitorPaywallListener(call)
            )
        } ?: run {
            call.reject("Activity not available")
        }
    }

    @PluginMethod
    fun presentPaywallIfNeeded(call: PluginCall) {
        val args = parsePaywallArgs(call)
        val requiredEntitlementIdentifier = args.requiredEntitlementIdentifier
        
        if (requiredEntitlementIdentifier == null) {
            call.reject("requiredEntitlementIdentifier is required for presentPaywallIfNeeded")
            return
        }
        
        activity?.let { activity ->
            PaywallActivityLauncher.launchIfNeeded(
                activity = activity,
                requiredEntitlementIdentifier = requiredEntitlementIdentifier,
                offeringIdentifier = args.offeringIdentifier,
                overrides = args.overrides,
                listener = CapacitorPaywallListener(call)
            )
        } ?: run {
            call.reject("Activity not available")
        }
    }

    @PluginMethod
    fun presentCustomerCenter(call: PluginCall) {
        activity?.let { activity ->
            CustomerCenterLauncher.launch(
                activity,
                listener = CapacitorCustomerCenterListener(call)
            )
        } ?: run {
            call.reject("Activity not available")
        }
    }

    @PluginMethod
    fun setPaywallOverrides(call: PluginCall) {
        val overridesObject = call.getObject("overrides") ?: run {
            call.reject("Missing overrides parameter")
            return
        }
        
        val overrides = parseOverrides(overridesObject)
        PaywallOverrides.setDefaultOverrides(overrides)
        call.resolve()
    }

    // MARK: - Helper Methods

    private fun parsePaywallArgs(call: PluginCall): PaywallArgs {
        val offeringIdentifier = call.getString("offeringIdentifier")
        val requiredEntitlementIdentifier = call.getString("requiredEntitlementIdentifier")
        
        var overrides: PaywallOverrides? = null
        call.getObject("overrides")?.let {
            overrides = parseOverrides(it)
        }
        
        return PaywallArgs(offeringIdentifier, requiredEntitlementIdentifier, overrides)
    }

    private fun parseOverrides(overridesObject: JSObject): PaywallOverrides {
        val overrides = PaywallOverrides()
        
        // Font options
        overridesObject.getString("fontFamily")?.let {
            overrides.fontFamily = it
        }
        
        // Colors
        overridesObject.getString("backgroundColor")?.let {
            overrides.paywallColors = overrides.paywallColors.copy(
                background = parseColor(it)
            )
        }
        
        overridesObject.getString("primaryTextColor")?.let {
            overrides.paywallColors = overrides.paywallColors.copy(
                primaryText = parseColor(it)
            )
        }
        
        overridesObject.getString("secondaryTextColor")?.let {
            overrides.paywallColors = overrides.paywallColors.copy(
                secondaryText = parseColor(it)
            )
        }
        
        overridesObject.getString("accentColor")?.let {
            overrides.paywallColors = overrides.paywallColors.copy(
                accent = parseColor(it)
            )
        }
        
        // Button options
        overridesObject.getString("primaryButtonBackgroundColor")?.let {
            overrides.paywallFooterColors = overrides.paywallFooterColors.copy(
                primaryButtonBackground = parseColor(it)
            )
        }
        
        overridesObject.getString("primaryButtonTextColor")?.let {
            overrides.paywallFooterColors = overrides.paywallFooterColors.copy(
                primaryButtonText = parseColor(it)
            )
        }
        
        overridesObject.getString("secondaryButtonBackgroundColor")?.let {
            overrides.paywallFooterColors = overrides.paywallFooterColors.copy(
                secondaryButtonBackground = parseColor(it)
            )
        }
        
        overridesObject.getString("secondaryButtonTextColor")?.let {
            overrides.paywallFooterColors = overrides.paywallFooterColors.copy(
                secondaryButtonText = parseColor(it)
            )
        }
        
        // Text overrides
        overridesObject.getString("title")?.let {
            overrides.paywallText = overrides.paywallText.copy(
                title = it
            )
        }
        
        overridesObject.getString("subtitle")?.let {
            overrides.paywallText = overrides.paywallText.copy(
                subtitle = it
            )
        }
        
        overridesObject.getString("callToAction")?.let {
            overrides.paywallText = overrides.paywallText.copy(
                callToAction = it
            )
        }
        
        overridesObject.getString("restoreButtonText")?.let {
            overrides.paywallText = overrides.paywallText.copy(
                restorePurchasesButtonText = it
            )
        }
        
        // Display options
        overridesObject.getBoolean("displayCloseButton")?.let {
            overrides.displayCloseButton = it
        }
        
        return overrides
    }

    private fun parseColor(colorString: String): PaywallColor {
        return try {
            val color = Color.parseColor(colorString)
            PaywallColor(color)
        } catch (e: Exception) {
            PaywallColor(Color.BLACK) // Default to black if invalid
        }
    }
}

// Helper class for passing Paywall arguments
private data class PaywallArgs(
    val offeringIdentifier: String?,
    val requiredEntitlementIdentifier: String?,
    val overrides: PaywallOverrides?
)

// Capacitor-specific Paywall listener implementation
private class CapacitorPaywallListener(private val call: PluginCall) : PaywallActivityLauncher.Listener {
    override fun onPaywallPresented() {
        // No need to notify for presentation
    }

    override fun onPaywallDismissed() {
        // No need to notify for dismissal
    }

    override fun onPurchaseStarted() {
        // No need to notify for purchase started
    }

    override fun onPurchaseCompleted(customerInfo: Any, storeTransaction: Any?) {
        val result = JSObject().apply {
            put("status", "purchased")
            if (storeTransaction != null) {
                put("transactionId", storeTransaction.toString())
            }
        }
        call.resolve(result)
    }

    override fun onPurchaseError(error: Throwable) {
        val result = JSObject().apply {
            put("status", "error")
            put("error", JSObject().apply {
                put("code", "purchase_error")
                put("message", error.message ?: "Unknown error")
            })
        }
        call.resolve(result)
    }

    override fun onRestoreStarted() {
        // No need to notify for restore started
    }

    override fun onRestoreCompleted(customerInfo: Any) {
        val result = JSObject().apply {
            put("status", "restored")
        }
        call.resolve(result)
    }

    override fun onRestoreError(error: Throwable) {
        val result = JSObject().apply {
            put("status", "error")
            put("error", JSObject().apply {
                put("code", "restore_error")
                put("message", error.message ?: "Unknown error")
            })
        }
        call.resolve(result)
    }

    override fun onPaywallError(error: Throwable) {
        val result = JSObject().apply {
            put("status", "error")
            put("error", JSObject().apply {
                put("code", "paywall_error")
                put("message", error.message ?: "Unknown error")
            })
        }
        call.resolve(result)
    }

    // Called when user cancels the paywall
    override fun onPaywallCancelled() {
        val result = JSObject().apply {
            put("status", "cancelled")
        }
        call.resolve(result)
    }
}

// Capacitor-specific Customer Center listener implementation
private class CapacitorCustomerCenterListener(private val call: PluginCall) : CustomerCenter.Listener {
    override fun onDismiss() {
        call.resolve()
    }

    override fun onError(error: Throwable) {
        call.reject(error.message ?: "Unknown error")
    }
} 